File: packages/zynth-ios/ios/ZynthKit/src/debug/PerformanceProfiler.swift
import Foundation

@objc
public final class PerformanceProfiler: NSObject {
    @objc public static let shared = PerformanceProfiler()

    @objc public var layoutStart: CFAbsoluteTime = 0
    @objc public var layoutEnd: CFAbsoluteTime = 0
    @objc public var renderStart: CFAbsoluteTime = 0
    @objc public var renderEnd: CFAbsoluteTime = 0

    private override init() {}

    @objc public func recordLayoutStart() {
        layoutStart = CFAbsoluteTimeGetCurrent()
    }

    @objc public func recordLayoutEnd() {
        layoutEnd = CFAbsoluteTimeGetCurrent()
    }

    @objc public func recordRenderStart() {
        renderStart = CFAbsoluteTimeGetCurrent()
    }

    @objc public func recordRenderEnd() {
        renderEnd = CFAbsoluteTimeGetCurrent()
    }

    @objc public func getFrameStats() -> [String: Any] {
        let layoutTime = (layoutEnd - layoutStart) * 1000
        let renderTime = (renderEnd - renderStart) * 1000
        return [
            "layoutTime": layoutTime,
            "renderTime": renderTime
        ]
    }
}

File: packages/zynth-android/android/ZynthKit/src/main/java/com/zynth/kit/debug/PerformanceProfiler.kt
package com.zynth.kit.debug

import java.util.concurrent.atomic.AtomicLong

object PerformanceProfiler {
    @Volatile
    var layoutStart: Long = 0
    @Volatile
    var layoutEnd: Long = 0
    @Volatile
    var renderStart: Long = 0
    @Volatile
    var renderEnd: Long = 0

    fun recordLayoutStart() {
        layoutStart = System.nanoTime()
    }

    fun recordLayoutEnd() {
        layoutEnd = System.nanoTime()
    }

    fun recordRenderStart() {
        renderStart = System.nanoTime()
    }

    fun recordRenderEnd() {
        renderEnd = System.nanoTime()
    }

    fun getFrameStats(): Map<String, Any> {
        val layoutTime = (layoutEnd - layoutStart) / 1_000_000.0
        val renderTime = (renderEnd - renderStart) / 1_000_000.0
        return mapOf(
            "layoutTime" to layoutTime,
            "renderTime" to renderTime
        )
    }
}
